src/shared/tasks.ts

```diff
@@ -0,0 +1,365 @@
+            state: 'stopped',
+            dispose: () => cancelToken.dispose(), // TODO: potentially force sweep on disposed tasks
+            cancel: cancelToken.cancel.bind(cancelToken),
+            isPending: () => task.state === 'pending',
+            isCompleted: () => task.state === 'completed',
+            isCancelled: () => task.state === 'cancelled',
+            start: (...args: Parameters<typeof fn>) => {
+                if ((task as PendingTask<T>).promise !== undefined) {
+                    return task
+                }
+
+                const throwIfCancelled = () => {
+                    if (cancelToken.isCancellationRequested) {
+                        setResult(Result.err(cancelToken.reason))
+                        this.updateInfo(task.id, { fulfilled: true, state: 'cancelled' })
+
+                        throw chainReason(cancelToken.reason)
+                    }
+                }
+
+                if (this.context) {
+                    if (this.context.cancelToken.isCancellationRequested) {
+                        const message = `Parent task "${this.context.taskId}" cancelled`
+                        cancelToken.cancel(new ToolkitError(message, { cause: this.context.cancelToken.reason }))
+                    }
+
+                    throwIfCancelled()
+                    this.link(task, this.context)
+                } else {
+                    throwIfCancelled()
+                    this.updateInfo(task.id, { state: 'pending' })
+                }
+
+                const ctx = { taskId: task.id, cancelToken }
+                const promise = this.#context
+                    .run(ctx, fn as (...args: any[]) => Promise<T>, ...args)
+                    .then(val => {
+                        setResult(Result.ok(val))
+                        throwIfCancelled()
+
+                        return val
+                    })
+                    .catch(err => {
+                        setResult(Result.err(err))
+
+                        if (err !== cancelToken.reason) {
+                            if (err instanceof CancellationError || (err instanceof ToolkitError && err.cancelled)) {
+                                cancelToken.cancel(err)
+                            } else {
+                                throwIfCancelled()
+                            }
+                        }
+
+                        throw err
+                    })
+                    .finally(() => {
+                        const finalState = task.state === 'pending' ? 'completed' : 'cancelled'
+                        this.updateInfo(task.id, { fulfilled: true, state: finalState })
+                        this.sweep(task.id)
+                    })
+
+                setPromise(promise)
+                return task
+            },
+        } as unknown as Task<T, U>
+
+        this.addTask(task)
+        return task as StoppedTask<T, U>
+    }
+
+    public getChildren(task: Pick<Task, 'id'>) {
```

Comment:
(maybe future) do we need a traverse() which visits all descendants?

Is there a `.then()` which "chains" a task to the completion of its parent? So a task can have many children, and then we can `.waitAll()` or `.waitAny()` on the children (or maybe descendants).

Reply:
> Is there a .then() which "chains" a task to the completion of its parent?

Right now child/parent relationship is strictly based off the asynchronous context so no. But I think it could make sense to add something like that because right now you have to nest things. Tasks can complete independently from one another which might be confusing at first but this is an accurate view of how the event loop works.

> So a task can have many children, and then we can .waitAll() or .waitAny() on the children (or maybe descendants).

I considered added more logic for these sorts of use-cases but decided to leave it as-is for now.